lib/Dialect/Arith/Conversions/ArithToModArith/ArithToModArith.cpp

```diff
@@ -75,6 +78,39 @@ struct ConvertConstant : public OpConversionPattern<mlir::arith::ConstantOp> {
   }
 };
 
+struct ConvertExt : public OpConversionPattern<mlir::arith::ExtSIOp> {
+  ConvertExt(mlir::MLIRContext *context)
+      : OpConversionPattern<mlir::arith::ExtSIOp>(context) {}
+
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      ::mlir::arith::ExtSIOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    ImplicitLocOpBuilder b(op.getLoc(), rewriter);
+
+    if (auto originOp = op.getOperand().getDefiningOp<memref::LoadOp>()) {
+      if (auto *globalOp = originOp.getMemRef().getDefiningOp()) {
```

Comment:
Thanks for the information. 

How would this separate pattern be implemented? I tried to write a separate OpConversionPattern for LoadOp, but that is only triggered it the function in addLegalOp would fail.

Reply:
Hmmm. This is a good point. I think you can actually make this work by adding a custom materialization in the TypeConverter via https://github.com/llvm/llvm-project/blob/dda1d1674755e0e68789e01ed8698ea91b0b54b0/mlir/include/mlir/Transforms/DialectConversion.h#L221

I _think_ it would be: a "target materialization" would insert mod_arith.encode, while a "source materialization" would insert mod_arith.extract, but I may have gotten those roles reversed. Basically, this would replace the default `unrealized_conversion_cast` when encountering a type error, with the appropriate op from `mod_arith`.

If you can't get this to work after trying for ten minutes, we can just submit it as is and maybe clean it up later.

Reply:
Implementation went actually pretty quickly (once I figured materialization out). 
Damn, this is a powerful and cool piece of MLIR. Thanks for the TIL.